### DISCUSSION TITLE
Nested Virtualization within Cloud Hypervisor

### DISCUSSION THREAD

**Discussion by ghost** [cloud-hypervisor/cloud-hypervisor#5194] [Q&A] [answered]
Is /dev/kvm available within the cloud hypervisor?

**Comment by michael2012z** [comment, marked as answer]
It depends on the architecture where you run the Cloud Hypervisor:
- Yes, if it's X86_64 (AFAIK)
- No, if it's AArch64.

**Reply by ghost** [reply]
I'm interested in x86_64 arch
Thank you @michael2012z

**Comment by alexellis** [comment]
@michael2012z I was also curious about this.

The Firecracker developers state that this is undefined behaviour and unsafe, and point to both qemu and cloud-hypervisor as examples of projects where this may be a supported use-case.

/dev/kvm can be used within Firecracker, however they advise against it when isolation is important

I couldn't find any documentation on CH stating that use of KVM within a VM is secure, can you advise?

https://github.com/firecracker-microvm/firecracker/issues/1721#issuecomment-1170278434

**Reply by rbradford** [reply]
The nested virtualisation in Cloud Hypervisor works the same way it works under KVM VMM. Principally the host KVM (which is why you need to enable nested support in the host KVM; most distributions turn that on anyway) will handle the exits and then "reflect" back to the guest in-between if necessary. 

Here is some relevant documentation how this works:

https://www.usenix.org/legacy/events/osdi10/tech/full_papers/Ben-Yehuda.pdf
https://events19.linuxfoundation.org/wp-content/uploads/2017/12/Improving-KVM-x86-Nested-Virtualization-Liran-Alon-Oracle.pdf

Ultimately you do need to trust every KVM implementation running between the hardware and the guest.

**Comment by michael2012z** [comment]
For AArch64, nested virtualization came around in Armv8.3-A, and an enhancement was added in Armv8.4-A. See this page: https://developer.arm.com/documentation/102378/0201/Armv8-x-and-Armv9-x-extensions-and-features

So the hardware (supporting at least v8.3) is the precondition.
Regarding the software, I am not quite sure about how the nested virtualization is supported in kernel (on AArch64).

On CH, we haven't tested and enabled the nested virtualization. One of the reasons is our CI server doesn't support it.

**Reply by alexellis** [reply]
@michael2012z if you're interested in CI that does support nested virtualisation, we did some experiments today with actuated and were able to run a Firecracker VM and QEMU/KVM VM for Nix. https://twitter.com/alexellisuk/status/1625468299176730624?s=20 / https://twitter.com/welteki/status/1625487015650410496?s=20

Thanks for the note on Arm - I'd understood that the current generation of Ampere servers available on Oracle Cloud / Google Cloud / Azure and (and AWS Graviton) are all on earlier versions of the architecture so don't support it yet.

**Reply by michael2012z** [reply]
Thank you for the sharing. Cool!
